### PR TITLE
Python: Fix core observability unsafe serialization of function-call arguments containing dataclass/framework objects

### LINKED ISSUE #5733: Python: Core observability cannot safely serialize workflow request_info payloads
## Description

PR `microsoft/agent-framework#3275` fixed the JSON serialization problem in the Agent Framework AG-UI integration, closing `microsoft/agent-framework#3239`, by serializing tool-call arguments with `make_json_safe()` before `json.dumps()`.

However, the same unsafe serialization pattern still appears to exist in `agent-framework-core` observability. In latest `agent-framework-core==1.3.0`, `agent_framework/observability.py::_capture_messages()` still serializes OpenTelemetry message payloads with raw `json.dumps(...)`, and the function-call branch keeps raw `content.arguments`.

I confirmed this is still an issue in the core workflow/request-info path: a workflow `request_info` event can be converted into function-call arguments containing a dataclass payload, and core observability then fails when `_capture_messages()` tries to serialize it with raw `json.dumps(...)`.

## Evidence from latest release

Package checked:

```text
agent-framework-core==1.3.0
agent-framework-foundry==1.3.0
agent-framework-foundry-hosting==1.0.0a260507
```

In `agent_framework/observability.py`:

```py
span.set_attribute(
    OtelAttr.OUTPUT_MESSAGES if output else OtelAttr.INPUT_MESSAGES,
    json.dumps(otel_messages, ensure_ascii=False),
)
```

The `function_call` branch preserves raw arguments:

```py
case "function_call":
    return {
        "type": "tool_call",
        "id": content.call_id,
        "name": content.name,
        "arguments": content.arguments,
    }
```

There is no `make_json_safe(...)` and no `default=...` encoder in this path.

## Why this matters

Workflow `request_info` events can contain framework or dataclass payloads. When those are surfaced as function-call arguments and telemetry captures messages, core observability can raise:

```text
TypeError: Object of type HandoffAgentUserRequest is not JSON serializable
```

This affects Handoff/workflow scenarios outside the AG-UI adapter path, including Foundry-hosted agent scenarios where telemetry captures messages.

## Expected behavior

Core observability should safely serialize message/tool-call payloads that contain dataclasses or framework objects emitted by workflow/handoff request-info flows.

## Actual behavior

Core observability still passes raw function-call arguments into `json.dumps(...)`, which can fail for dataclass/framework payloads.

## Related issues / PRs

- `microsoft/agent-framework#3239` reported the AG-UI serialization failure.
- `microsoft/agent-framework#3275` fixed the AG-UI path by serializing arguments with `make_json_safe(content.arguments)`.
- This issue is about the remaining core observability path in `agent-framework-core`.

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

Core observability in `agent_framework/observability.py` passed raw `function_call` arguments directly into `json.dumps`, which raises `TypeError` when those arguments contain dataclass or framework objects (e.g. `HandoffAgentUserRequest`) emitted by workflow `request_info` / handoff flows. This affected any Foundry-hosted or handoff scenario with sensitive telemetry enabled.

Fixes #5733

### Description

The root cause was that `_to_otel_part`'s `function_call` branch kept `content.arguments` as-is, with no safe-serialization step, mirroring the already-fixed AG-UI path. The fix lifts a `make_json_safe()` helper into `agent_framework/_serialization.py` — which recursively converts dataclasses, Pydantic models, datetimes, and `__dict__`-bearing objects to JSON-serializable primitives — and calls it on `content.arguments` before the value is included in the OTel span attribute. Tests covering dataclass payloads and nested non-primitive objects in `function_call` arguments were added to `test_observability.py` to prevent regression.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

---

> Note: PR autogenerated by moonbox3's agent

### CODE REVIEW

**Comment by moonbox3** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py">_serialization.py</a></td><td>150</td><td>15</td><td>90%</td><td><a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L517">517</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L534">534</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L544">544</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L612">612</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L637">637</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L646-L649">646&ndash;649</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L651-L654">651&ndash;654</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L658">658</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_serialization.py#L661">661</a></td></tr><tr><td colspan="5"><b>packages/core/agent_framework/_workflows</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py">_agent.py</a></td><td>363</td><td>64</td><td>82%</td><td><a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L68">68</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L76-L82">76&ndash;82</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L118-L119">118&ndash;119</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L212">212</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L273">273</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L286">286</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L353">353</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L364">364</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L366">366</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L426">426</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L432">432</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L449-L450">449&ndash;450</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L457">457</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L459">459</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L465">465</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L531-L532">531&ndash;532</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L541">541</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L580">580</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L615">615</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L690">690</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L720">720</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L737">737</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L765">765</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L776-L779">776&ndash;779</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L785">785</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L791">791</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L795-L796">795&ndash;796</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L799-L805">799&ndash;805</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L809-L810">809&ndash;810</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L818">818</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L879">879</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L886">886</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L892-L893">892&ndash;893</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L904">904</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L936">936</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L943">943</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L964">964</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L973">973</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L977">977</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L979-L981">979&ndash;981</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_agent.py#L988">988</a></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py">_functional.py</a></td><td>491</td><td>31</td><td>93%</td><td><a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L398-L399">398&ndash;399</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L465">465</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L488">488</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L497-L499">497&ndash;499</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L502">502</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L532">532</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L713-L714">713&ndash;714</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L721">721</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L969">969</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1001">1001</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1004-L1005">1004&ndash;1005</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1013-L1015">1013&ndash;1015</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1090-L1091">1090&ndash;1091</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1097">1097</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1480-L1481">1480&ndash;1481</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1483">1483</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1490-L1493">1490&ndash;1493</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1514">1514</a>, <a href="https://github.com/microsoft/agent-framework/blob/c7f4d9c7dd82c58895054676ec6fa1297951e032/packages/core/agent_framework/_workflows/_functional.py#L1539">1539</a></td></tr><tr><td><b>TOTAL</b></td><td><b>37429</b></td><td><b>4354</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 7458 | 34 :zzz: | 0 :x: | 0 :fire: | 1m 58s :stopwatch: |

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 2 | **Confidence:** 92% | **Result:** All clear

Reviewed: Security Reliability, Design Approach

---
*Automated review by moonbox3's agents*

File: python/packages/core/agent_framework/_serialization.py [hunk 1]
```diff
@@ -614,3 +616,46 @@ def _get_type_identifier(cls, value: Mapping[str, Any] | None = None) -> str:
...
+            return make_json_safe(obj.to_dict())  # type: ignore[no-any-return]
+        except TypeError:
+            pass
+    if callable(getattr(obj, "dict", None)):
+        try:
+            return make_json_safe(obj.dict())  # type: ignore[no-any-return]
+        except TypeError:
+            pass
+    if isinstance(obj, dict):
+        return {str(key): make_json_safe(value) for key, value in obj.items()}  # type: ignore[misc]
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_serialization.py:656]: Silent key collision: if the dict contains both `42` (int) and `"42"` (str), `str(42) == "42"` so one value silently overwrites the other with no warning. The new tests only cover non-overlapping keys. Consider logging a warning on collision, or documenting that key-uniqueness after string coercion is the caller's responsibility.

```suggestion
        coerced: dict = {}
        for key, value in obj.items():
            str_key = str(key)
            coerced[str_key] = make_json_safe(value)  # last writer wins on collision
        return coerced  # type: ignore[misc]
```

File: python/packages/core/tests/core/test_observability.py [hunk 2]
```diff
@@ -1691,6 +1691,94 @@ def test_to_otel_part_function_call():
...
+
+    from agent_framework._serialization import make_json_safe
+
+    class ObjWithNonCallableModelDump:
+        model_dump = 42  # not callable
+
+    obj = ObjWithNonCallableModelDump()
+    result = make_json_safe(obj)
+    # Falls back to __dict__ or str — must not raise TypeError
+    assert json.dumps(result) is not None
```
Comment (moonbox3) [inline python/packages/core/tests/core/test_observability.py]: `json.dumps(result) is not None` is trivially true — `json.dumps` always returns a string. This does not verify the correct fallback value. Since `ObjWithNonCallableModelDump` has no instance attributes, `vars(obj)` is `{}`, so `make_json_safe` should return `{}`. Assert that directly.

```suggestion
    assert result == {}
```

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 4 | **Confidence:** 86%

### ✓ Correctness
> The PR correctly fixes two bugs: (1) `hasattr` checks replaced with `callable(getattr(..., None))` to skip non-callable attributes named `model_dump`/`to_dict`/`dict`, and (2) dict keys are now stringified so `json.dumps` never raises `TypeError` on non-string keys. The `try/except TypeError` wrappers provide a safe fallback chain. The `__dict__` branch retains bare `key` references but `vars(obj)` always yields string keys, so no issue there. No correctness problems found.

### ✓ Security Reliability
> The PR correctly fixes two bugs: (1) using `callable(getattr(...))` to guard against non-callable attributes named `model_dump`/`to_dict`/`dict`, and (2) converting dict keys to strings so `json.dumps` never raises `TypeError` on non-string keys. Both fixes are well-motivated and tested. One reliability concern: the `str(key)` conversion at line 656 can silently clobber data when a dict contains both an integer key and its string equivalent (e.g., `{42: 'a', '42': 'b'}`), since iteration order determines which value survives with no warning. This is a silent data-loss scenario not covered by the new tests. The `TypeError` suppression around serialization method calls is reasonable in scope since `TypeError` on a zero-argument call is a narrow signal, but it could mask bugs in `model_dump`/`to_dict` implementations that raise `TypeError` internally — the fallback to `__dict__` may then silently serialize incorrect/partial state.

### ✓ Test Coverage
> The fix correctly guards against non-callable `model_dump`/`to_dict`/`dict` attributes and converts dict keys to strings. Two test coverage gaps: (1) the assertion in `test_make_json_safe_non_callable_method_attribute` is trivially true (`json.dumps` never returns `None`), masking whether the correct fallback value is returned; (2) the new `try/except TypeError` path — where a callable `model_dump()` raises `TypeError` and falls through — is never exercised by a test.

### ✓ Design Approach
> I did not find a design-approach issue in this bounded pass. The change stays focused on making observability serialization tolerate the two concrete edge cases covered by the diff: non-callable serializer-like attributes and non-string dict keys that would otherwise break JSON encoding.

---
*Automated review by moonbox3's agents*

File: python/packages/core/agent_framework/_serialization.py [hunk 3]
```diff
@@ -614,3 +616,37 @@ def _get_type_identifier(cls, value: Mapping[str, Any] | None = None) -> str:
...
+    if isinstance(obj, (datetime, date)):
+        return obj.isoformat()
+    if is_dataclass(obj) and not isinstance(obj, type):
+        return make_json_safe(asdict(obj))  # type: ignore[arg-type]
+    if hasattr(obj, "model_dump"):
+        return make_json_safe(obj.model_dump())  # type: ignore[no-any-return]
+    if hasattr(obj, "to_dict"):
+        return make_json_safe(obj.to_dict())  # type: ignore[no-any-return]
+    if hasattr(obj, "dict"):
+        return make_json_safe(obj.dict())  # type: ignore[no-any-return]
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/core/agent_framework/_serialization.py]: Fixed in `b8211136`.

File: python/packages/core/agent_framework/_serialization.py [hunk 4]
```diff
@@ -614,3 +616,37 @@ def _get_type_identifier(cls, value: Mapping[str, Any] | None = None) -> str:
...
+    if isinstance(obj, dict):
+        return {key: make_json_safe(value) for key, value in obj.items()}  # type: ignore[misc]
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/core/agent_framework/_serialization.py]: Fixed in `b8211136`.

File: python/packages/core/agent_framework/_serialization.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_serialization.py:656]: Fixed in `b8211136`.

File: python/packages/core/tests/core/test_observability.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/core/tests/core/test_observability.py]: Fixed in `b8211136`.

File: python/packages/core/agent_framework/observability.py [hunk 5]
```diff
@@ -2229,7 +2229,14 @@ def _to_otel_part(content: Content) -> dict[str, Any] | None:
...
             }
         case "function_call":
-            return {"type": "tool_call", "id": content.call_id, "name": content.name, "arguments": content.arguments}
+            from ._serialization import make_json_safe  # pyright: ignore[reportPrivateUsage]
+
+            return {
+                "type": "tool_call",
+                "id": content.call_id,
+                "name": content.name,
+                "arguments": make_json_safe(content.arguments),
```
Comment (eavanvalkenburg) [inline python/packages/core/agent_framework/observability.py]: we need to do something smarter here, because this will slow down as this is called every turn, this was one of the reasons things were slow at some point and that was why the logic was pushed to writing the Content and making sure that that is serializable, if that is not the case for some workflow setups, then let's fix that instead of having to do a lot of work every turn, this PR made that change: https://github.com/microsoft/agent-framework/pull/3014

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]

**Review by eavanvalkenburg — APPROVED** [review]

**Review by giles17 — APPROVED** [review]
